### PR TITLE
Adds support for per metric-type expiry intervals.

### LINKED ISSUE #300: Per metric-type expiry interval
How I can send to back the previous value, if the gauge is not updated at the next flush?

### PR DESCRIPTION
**Pull request by tiedotguy**
Closes #300.

### CODE REVIEW

File: pkg/statsd/aggregator.go [hunk 1]
```diff
@@ -187,8 +202,8 @@ func (a *MetricAggregator) Process(f ProcessFunc) {
 	f(a.metricMap)
 }
 
-func (a *MetricAggregator) isExpired(now, ts gostatsd.Nanotime) bool {
-	return a.expiryInterval != 0 && time.Duration(now-ts) > a.expiryInterval
+func (a *MetricAggregator) isExpired(interval time.Duration, now, ts gostatsd.Nanotime) bool {
```
Comment (ash2k) [inline pkg/statsd/aggregator.go:205]: Can be a plain function rather than a method.

File: pkg/statsd/statsd.go [hunk 2]
```diff
@@ -8,14 +8,15 @@ import (
...
-	"github.com/atlassian/gostatsd/pkg/web"
 	"github.com/libp2p/go-reuseport"
 	log "github.com/sirupsen/logrus"
 	"github.com/spf13/viper"
 	"golang.org/x/time/rate"
+
+	"github.com/atlassian/gostatsd"
+	"github.com/atlassian/gostatsd/pkg/stats"
+	"github.com/atlassian/gostatsd/pkg/transport"
+	"github.com/atlassian/gostatsd/pkg/web"
```
Comment (ash2k) [inline pkg/statsd/statsd.go:19]: What is happening with formatting? :)

File: pkg/statsd/aggregator.go (hunk 1 above)
Comment (tiedotguy) [reply pkg/statsd/aggregator.go]: Good catch, fixed

File: pkg/statsd/statsd.go (hunk 2 above)
Comment (tiedotguy) [reply pkg/statsd/statsd.go:19]: I think goland did a format, and it should be re-ordered anyway.

**Comment by MovieStoreGuy** [comment]
Do we know what the impact to memory would be adding this in?

**Comment by tiedotguy** [comment]
> Do we know what the impact to memory would be adding this in?

For space, effectively nothing (24 bytes * number of aggregators, which is the number of logical procs by default). `MetricAggregator`s are long lived structures created on startup, not a per-metric or per-aggregated-metric structure.

For bus / cache pressure, also effectively nothing. It's not the most efficient because it's 8->32 bytes per aggregator (it could be a single shared data structure for instance), however it's also only used on flush, which isn't the hot path.

**Pull request merged**